engine/include/cubos/engine/ui/canvas/draw_list.hpp

```diff
@@ -0,0 +1,74 @@
+#include <vector>
+
+#include <cubos/core/gl/render_device.hpp>
+
+#include <cubos/engine/ui/canvas/draw_command.hpp>
+#include <cubos/engine/ui/canvas/draw_type.hpp>
+
+namespace cubos::engine
+{
+    /// @brief Describes a draw commands and how it fits into the draw list carrying it.
+    /// @ingroup ui-canvas-plugin
+    struct UIDrawListEntry
+    {
+        UIDrawType& type;
+        UIDrawCommand command;
+        size_t offset;
+    };
+
+    /// @brief Holds a collections of draw commands and their data.
+    /// @ingroup ui-canvas-plugin
+    class UIDrawList
+    {
+    public:
+        /// @brief Constructs a draw list.
+        UIDrawList();
+
+        /// @brief Move constructs.
+        /// @param other Other draw list.
+        UIDrawList(UIDrawList&& other) noexcept
+            : entries(other.entries)
+            , perElementData(other.perElementData)
+        {
+            other.perElementData = nullptr;
+        }
+
+        /// @brief Copy constructs.
+        /// @param other Other draw list.
+        UIDrawList(const UIDrawList& other) noexcept
+            : entries(other.entries)
+        {
+            perElementData = operator new(sizeof(char) * 1024);
+            std::memcpy(perElementData, other.perElementData, sizeof(char) * 1024);
+        }
+        ~UIDrawList()
+        {
+            operator delete(perElementData);
+        }
+
+        /// @brief Adds entry to draw list.
+        /// @param entry Entry to add.
+        /// @param data Pointer to per element data.
+        void pushEntry(UIDrawListEntry entry, const void* data);
+
+        /// @brief Creates new draw command and adds it to draw list.
+        /// @param type Type of the command.
+        /// @param vertexArray Vertex array handle.
+        /// @param vertexOffset Index of the first vertex to be drawn.
+        /// @param vertexCount Number of vertexes to draw.
+        /// @param data Pointer to per element data.
+        /// @return The created command.
+        UIDrawCommand& emplaceCommand(UIDrawType& type, core::gl::VertexArray vertexArray, size_t vertexOffset,
+                                      size_t vertexCount, void* data);
+
+        /// @brief Empties draw list.
+        void clear();
+
+        std::vector<UIDrawListEntry> entries; ///< List holding the draw commands of this element and all its children.
+
+        void* perElementData; ///< Buffer holding the data for all the draw commands.
```

Comment:
These should be private, if they're not meant to be modified directly.
If they need to be accessed, add `const` getters:

```cpp
std::size_t size() const;
const UIDrawListEntry& entry(std::size_t i) const;
const void* data(std::size_t i) const;
```